### PR TITLE
Support `getitem` with labels for Series or column names for DataFrame

### LINKED ISSUE #625: Support selecting out lower-dimensional slices with [] for DataFrame and series
**Is your feature request related to a problem? Please describe.**

Support basic indexing for DataFrame.

Object Type | Selection | Return Value Type
-- | -- | --
Series | series[label] | scalar value
DataFrame | frame[colname] | Series corresponding to colname

### PR DESCRIPTION
**Pull request by hekaisheng**
## What do these changes do?

Support selecting out lower-dimensional slices with [] for DataFrame and series.

- [x] This PR may conflict with #638 , should be rebased after #638 is merged.

- [x] Add some cases for series and support getitem for non range index series.

This pull request is not a complete implementation of `__getitem__`, we should support selecting by slices or a boolean array(both `ndarray` and `mars.tensor`).

## Related issue number

Resolve #625

### CODE REVIEW

File: mars/dataframe/indexing/getitem.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    def __init__(self, col_names=None, object_type=ObjectType.series, **kw):
+        super(DataFrameIndex, self).__init__(_col_names=col_names, _object_type=object_type, **kw)
+
+    @property
+    def col_names(self):
+        return self._col_names
+
+    def __call__(self, df):
+        # if col_names is a tuple or list, return a DataFrame, else return a Series
+        if isinstance(self._col_names, (list, tuple)):
```
Comment (sighingnow) [inline mars/dataframe/indexing/getitem.py]: The `_col_names` won't be `tuple` (to be consistent with pandas).

File: mars/dataframe/indexing/getitem.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+                column_index = calc_columns_index(col_name, in_df)
+                selected_index.append(column_index)
+
+            # combine columns in one chunk and keep the columns order at the same time.
+            column_splits = []
+            start = 0
+            for i in range(len(selected_index)):
+                if i == len(selected_index) - 1 or selected_index[i] != selected_index[i + 1]:
+                    column_splits.append((col_names[slice(start, i + 1)], selected_index[start]))
+                    start = i + 1
```
Comment (sighingnow) [inline mars/dataframe/indexing/getitem.py]: This part looks really complicated... How to about grouping the neighboring column names like this:

```python
selected_index = [calc_columns_index(col, in_df) for col in col_names]
column_splits = np.split(col_names, np.where(np.diff(selected_index))[0] + 1)
column_indexes = np.split(selected_index, np.where(np.diff(selected_index))[0] + 1)
```

File: mars/dataframe/indexing/getitem.py (hunk 1 above)
Comment (hekaisheng) [reply mars/dataframe/indexing/getitem.py]: Fixed.

**Comment by hekaisheng** [comment]
This pull request is not a complete implementation of `__getitem__`, we should support selecting by slices or a boolean array(both `ndarray` and `mars.tensor`) in the future.

File: mars/dataframe/indexing/getitem.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+                if len(chks) == 1:
+                    chk = chks[0]
+                else:
+                    chk_op = SeriesIndex(labels=op.labels)
+                    chk = chk_op.new_chunk(chks, shape=(np.nan,), dtype=chks[0].dtype,
+                                           index_value=parse_index(pd.RangeIndex(0)))
+                new_chunks.append(chk)
+            chunks = new_chunks
+
+        if isinstance(op.labels, list):
```
Comment (qinxuye) [inline mars/dataframe/indexing/getitem.py]: Add comment here.

File: mars/dataframe/indexing/getitem.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+            index_op = SeriesIndex(labels=op.labels, is_terminal=True)
+            chunk = index_op.new_chunk(chunks, shape=(len(op.labels),), dtype=chunks[0].dtype,
+                                       index_value=parse_index(pd.Index(op.labels)))
+            new_op = op.copy()
+            return new_op.new_seriess(op.inputs, shape=out_series.shape, chunks=[chunk], dtype=out_series.dtype,
+                                      index_value=out_series.index_value, nsplits=((len(op.labels),),))
+        else:
```
Comment (qinxuye) [inline mars/dataframe/indexing/getitem.py]: Ditto.

File: mars/dataframe/indexing/getitem.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+            new_op = op.copy()
+            return new_op.new_scalars(op.inputs, dtype=out_series.dtype, chunks=[chunk])
+
+    @classmethod
+    def tile(cls, op):
+        in_series = op.inputs[0]
+        out_series = op.outputs[0]
+        if len(in_series.chunks) == 1:
+            index_op = SeriesIndex(labels=op.labels, is_terminal=True)
+            if isinstance(op.labels, list):
```
Comment (qinxuye) [inline mars/dataframe/indexing/getitem.py]: Logic like this is everywhere, could you think a better way to reduce it?

File: mars/dataframe/indexing/getitem.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+from .utils import calc_columns_index
+
+
+class SeriesIndex(DataFrameOperand, DataFrameOperandMixin):
+    _op_type_ = OperandDef.INDEX
+
+    _labels = AnyField('labels')
+
+    _combine_size = Int32Field('combine_size')
+    _is_intermediate = BoolField('is_terminal')
```
Comment (qinxuye) [inline mars/dataframe/indexing/getitem.py]: is_terminal -> is_intermediate

**Pull request merged**